Nesibe Elibol

# Project Teil 1
## DataPreprocessing


In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

## Einlesen der Daten 

In [2]:
df1 = pd.read_csv("part1.csv", 
                       sep=',', 
                       engine='python',   # engine='python' da sonst Warning erhält
                       nrows=2)           # nur die ersten 2 Zeilen einlesen, um zu schauen wie der sep aussieht
df1 

Unnamed: 0  funded_amount  loan_amount             activity  \
0           0          300.0        300.0  Fruits & Vegetables   
1           1          575.0        575.0             Rickshaw   

           sector                                                use  \
0            Food            To buy seasonal, fresh fruits to sell.    
1  Transportation  to repair and maintain the auto rickshaw used ...   

  country_code   country  region currency  term_in_months  lender_count  \
0           PK  Pakistan  Lahore      PKR            12.0            12   
1           PK  Pakistan  Lahore      PKR            11.0            14   

  borrower_genders repayment_interval  
0           female          irregular  
1   female, female          irregular

In [3]:
df2 = pd.read_csv("part2.csv", 
                       sep='/n', 
                       engine='python',   
                       nrows=2)           
df2 

#funded_amount#loan_amount#activity#sector#use#country_code#country#region#currency#term_in_months#lender_count#borrower_genders#repayment_interval
0  0#175.0#175.0#Liquor Store / Off-License#Food#...                                                                                                 
1  1#325.0#325.0#Livestock#Agriculture#to buy 3 z...

In [4]:
#Datensätze vollständig einlesen und Prüfen 


df1=pd.read_csv("part1.csv",sep=',') 
df2=pd.read_csv("part2.csv",sep='#')  


df1.head()


Unnamed: 0  funded_amount  loan_amount             activity  \
0           0          300.0        300.0  Fruits & Vegetables   
1           1          575.0        575.0             Rickshaw   
2           2          150.0        150.0       Transportation   
3           3          200.0        200.0           Embroidery   
4           4          400.0        400.0           Milk Sales   

           sector                                                use  \
0            Food            To buy seasonal, fresh fruits to sell.    
1  Transportation  to repair and maintain the auto rickshaw used ...   
2  Transportation  To repair their old cycle-van and buy another ...   
3            Arts  to purchase an embroidery machine and a variet...   
4            Food                           to purchase one buffalo.   

  country_code   country        region currency  term_in_months  lender_count  \
0           PK  Pakistan        Lahore      PKR            12.0            12   
1           PK  Pakistan        Lahore      PKR            11.0            14   
2           IN     India     Maynaguri      INR            43.0             6   
3           PK  Pakistan        Lahore      PKR            11.0             8   
4           PK  Pakistan  Abdul Hakeem      PKR            14.0            16   

  borrower_genders repayment_interval  
0           female          irregular  
1   female, female          irregular  
2           female             bullet  
3           female          irregular  
4           female            monthly

In [5]:
df2.head()

Unnamed: 0  funded_amount  loan_amount                    activity  \
0           0          175.0        175.0  Liquor Store / Off-License   
1           1          325.0        325.0                   Livestock   
2           2          550.0        550.0                  Food Stall   
3           3         1300.0       1300.0                      Cattle   
4           4          900.0        900.0              Consumer Goods   

         sector                                                use  \
0          Food  to purchase additional stock of coconut wine t...   
1   Agriculture         to buy 3 zebus and food to fatten them up.   
2          Food  to buy ingredients for her food-vending busine...   
3   Agriculture                         to buy one head of cattle.   
4  Personal Use              to buy consumer goods amongst others.   

  country_code      country                       region currency  \
0           PH  Philippines                  Palo, Leyte      PHP   
1           MG   Madagascar                    Antsirabe      MGA   
2           PH  Philippines                Cordova, Cebu      PHP   
3           EG        Egypt                     Baniswef      EGP   
4           PE         Peru  Urubamba - Urubamba - Cusco      PEN   

   term_in_months  lender_count borrower_genders repayment_interval  
0             8.0             6           female          irregular  
1            12.0            13           female            monthly  
2             5.0             6           female          irregular  
3            14.0            50             male            monthly  
4             6.0             1           female          irregular

## Verbinden der DataFrames

In [6]:
#Dataframes verbinden und Prüfen

df=pd.concat([df1, df2])
df.head()

Unnamed: 0  funded_amount  loan_amount             activity  \
0           0          300.0        300.0  Fruits & Vegetables   
1           1          575.0        575.0             Rickshaw   
2           2          150.0        150.0       Transportation   
3           3          200.0        200.0           Embroidery   
4           4          400.0        400.0           Milk Sales   

           sector                                                use  \
0            Food            To buy seasonal, fresh fruits to sell.    
1  Transportation  to repair and maintain the auto rickshaw used ...   
2  Transportation  To repair their old cycle-van and buy another ...   
3            Arts  to purchase an embroidery machine and a variet...   
4            Food                           to purchase one buffalo.   

  country_code   country        region currency  term_in_months  lender_count  \
0           PK  Pakistan        Lahore      PKR            12.0            12   
1           PK  Pakistan        Lahore      PKR            11.0            14   
2           IN     India     Maynaguri      INR            43.0             6   
3           PK  Pakistan        Lahore      PKR            11.0             8   
4           PK  Pakistan  Abdul Hakeem      PKR            14.0            16   

  borrower_genders repayment_interval  
0           female          irregular  
1   female, female          irregular  
2           female             bullet  
3           female          irregular  
4           female            monthly

In [7]:
#Index resetten
df.reset_index(inplace=True)
df.head()

index  Unnamed: 0  funded_amount  loan_amount             activity  \
0      0           0          300.0        300.0  Fruits & Vegetables   
1      1           1          575.0        575.0             Rickshaw   
2      2           2          150.0        150.0       Transportation   
3      3           3          200.0        200.0           Embroidery   
4      4           4          400.0        400.0           Milk Sales   

           sector                                                use  \
0            Food            To buy seasonal, fresh fruits to sell.    
1  Transportation  to repair and maintain the auto rickshaw used ...   
2  Transportation  To repair their old cycle-van and buy another ...   
3            Arts  to purchase an embroidery machine and a variet...   
4            Food                           to purchase one buffalo.   

  country_code   country        region currency  term_in_months  lender_count  \
0           PK  Pakistan        Lahore      PKR            12.0            12   
1           PK  Pakistan        Lahore      PKR            11.0            14   
2           IN     India     Maynaguri      INR            43.0             6   
3           PK  Pakistan        Lahore      PKR            11.0             8   
4           PK  Pakistan  Abdul Hakeem      PKR            14.0            16   

  borrower_genders repayment_interval  
0           female          irregular  
1   female, female          irregular  
2           female             bullet  
3           female          irregular  
4           female            monthly

In [8]:
#Entfernen der Spalten Unnamed:0 
df.drop(columns=["Unnamed: 0","index"],inplace=True)
df.head()

funded_amount  loan_amount             activity          sector  \
0          300.0        300.0  Fruits & Vegetables            Food   
1          575.0        575.0             Rickshaw  Transportation   
2          150.0        150.0       Transportation  Transportation   
3          200.0        200.0           Embroidery            Arts   
4          400.0        400.0           Milk Sales            Food   

                                                 use country_code   country  \
0            To buy seasonal, fresh fruits to sell.            PK  Pakistan   
1  to repair and maintain the auto rickshaw used ...           PK  Pakistan   
2  To repair their old cycle-van and buy another ...           IN     India   
3  to purchase an embroidery machine and a variet...           PK  Pakistan   
4                           to purchase one buffalo.           PK  Pakistan   

         region currency  term_in_months  lender_count borrower_genders  \
0        Lahore      PKR            12.0            12           female   
1        Lahore      PKR            11.0            14   female, female   
2     Maynaguri      INR            43.0             6           female   
3        Lahore      PKR            11.0             8           female   
4  Abdul Hakeem      PKR            14.0            16           female   

  repayment_interval  
0          irregular  
1          irregular  
2             bullet  
3          irregular  
4            monthly

## Vertraut werden mit den Daten

In [9]:
#Betrachtung der Lagemaße und Erkennung von Extremwerten
df.describe()

funded_amount    loan_amount  term_in_months   lender_count
count  671205.000000  671205.000000   671205.000000  671205.000000
mean      785.995061     842.397107       13.739022      20.590922
std      1130.398941    1198.660073        8.598919      28.459551
min         0.000000      25.000000        1.000000       0.000000
25%       250.000000     275.000000        8.000000       7.000000
50%       450.000000     500.000000       13.000000      13.000000
75%       900.000000    1000.000000       14.000000      24.000000
max    100000.000000  100000.000000      158.000000    2986.000000

In [10]:
#Datentypen der Spalten 
df.dtypes

funded_amount         float64
loan_amount           float64
activity               object
sector                 object
use                    object
country_code           object
country                object
region                 object
currency               object
term_in_months        float64
lender_count            int64
borrower_genders       object
repayment_interval     object
dtype: object

In [11]:
#Spaltennamen wiedergeben 
df.columns

Index(['funded_amount', 'loan_amount', 'activity', 'sector', 'use',
       'country_code', 'country', 'region', 'currency', 'term_in_months',
       'lender_count', 'borrower_genders', 'repayment_interval'],
      dtype='object')

In [12]:
#Anzahl der unique Werte je Spalte anschauen
df.nunique()

funded_amount            610
loan_amount              479
activity                 163
sector                    15
use                   424912
country_code              86
country                   87
region                 12695
currency                  67
term_in_months           148
lender_count             503
borrower_genders       11298
repayment_interval         4
dtype: int64

In [13]:
#Mögliche Duplikate oder auf Missing Synonymen achten
df.sector.unique()

array(['Food', 'Transportation', 'Arts', 'Services', 'Agriculture',
       'Manufacturing', 'Wholesale', 'Retail', 'Clothing', 'Construction',
       'Health', 'Education', 'Personal Use', 'Housing', 'Entertainment'],
      dtype=object)

In [14]:
#weitere Informationen holen
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 671205 entries, 0 to 671204
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   funded_amount       671205 non-null  float64
 1   loan_amount         671205 non-null  float64
 2   activity            671205 non-null  object 
 3   sector              671205 non-null  object 
 4   use                 666973 non-null  object 
 5   country_code        671197 non-null  object 
 6   country             671205 non-null  object 
 7   region              614405 non-null  object 
 8   currency            671205 non-null  object 
 9   term_in_months      671205 non-null  float64
 10  lender_count        671205 non-null  int64  
 11  borrower_genders    666984 non-null  object 
 12  repayment_interval  671205 non-null  object 
dtypes: float64(3), int64(1), object(9)
memory usage: 66.6+ MB


In [15]:
#Betrachtung der Unique Werte in der Spalte borrower_genders
df.borrower_genders.unique()

array(['female', 'female, female', 'female, female, female', ...,
       'female, female, male, female, female, female, female, female, female, female, male, male, female, female, male, female, female, female, female, female, female, female',
       'male, female, female, female, female, female, female, female, male, male, female, male, female, male, male, male',
       'female, female, female, male, female, female, female, male, female, female, female, male, female, male, female, female, female, female, female, female, female, female, female, female, female, female, female, female, male'],
      dtype=object)

## Fehlende Werte

### Konventionen suchen

In [16]:
#Betrachtung der Konventionen von fehlenden Werten, wie null, nan  
df.isnull().sum()

funded_amount             0
loan_amount               0
activity                  0
sector                    0
use                    4232
country_code              8
country                   0
region                56800
currency                  0
term_in_months            0
lender_count              0
borrower_genders       4221
repayment_interval        0
dtype: int64

In [17]:
df.isna().sum()/(len(df))*100

funded_amount         0.000000
loan_amount           0.000000
activity              0.000000
sector                0.000000
use                   0.630508
country_code          0.001192
country               0.000000
region                8.462392
currency              0.000000
term_in_months        0.000000
lender_count          0.000000
borrower_genders      0.628869
repayment_interval    0.000000
dtype: float64

Die Spalte "use" enthält kurze Beschreibung über die Geldverwendung in Projekten. Die Spalten sektor und activity beschreiben und grenzen die Projekte bereits thematisch ein, sodass zunächst "use" bei einem Datensatz mit einer Größe von über 670.000 Zeilen nicht weiter betrachten werden kann. Entsprechend wird diese entfernt.

Die Spalte Region enthält die Regionen aus denen die Projekte stammen. Mit 8.46% ist das der größte Wert der fehlenden Daten und wird behalten. Weitere Marktanalyse über die können detaillierte Insights über den lokalen Markt geben und verschafft eine interessante Entwicklung der Regionen über Jahre. Durch einen potentiellen Mehrwert in der Visuallisierung wreden die Missings nicht entfernt. 

Die Spalte borrower genders wird behalten, aber da aus borrower genders ein Feature entstehen soll, werden die Missings gelöscht. Das macht bei 4221 Zeilen mit fehlenden Werten zu 671.205 Zeilen einen marginalen Anteil von 0,63%. Die Betrachtung von 99,37% der Daten ergibt immer noch eine aussagekräftige Auswertung. 

Die Spalte Country_Code wird ergänzt. 

In [18]:
df.drop(columns=["use"],inplace=True) 

In [19]:
lis=df.loc[df.borrower_genders.isnull()].index
df.drop(lis,inplace=True)
df.isnull().sum()

funded_amount             0
loan_amount               0
activity                  0
sector                    0
country_code              8
country                   0
region                52579
currency                  0
term_in_months            0
lender_count              0
borrower_genders          0
repayment_interval        0
dtype: int64

In [20]:
#Betrachtung der Missing in country_code, ausschlißelich Namibia betroffen
df.loc[df.loc[:,'country_code'].isnull(), ['country_code','country'] ]

country_code  country
202537          NaN  Namibia
202823          NaN  Namibia
344929          NaN  Namibia
351177          NaN  Namibia
420953          NaN  Namibia
421218          NaN  Namibia
487207          NaN  Namibia
487653          NaN  Namibia

In [21]:
df.loc[:,'country_code'].fillna('NA', inplace=True)

In [22]:
#Prüfen
df.isnull().sum()

funded_amount             0
loan_amount               0
activity                  0
sector                    0
country_code              0
country                   0
region                52579
currency                  0
term_in_months            0
lender_count              0
borrower_genders          0
repayment_interval        0
dtype: int64

In [23]:
df.head()

funded_amount  loan_amount             activity          sector  \
0          300.0        300.0  Fruits & Vegetables            Food   
1          575.0        575.0             Rickshaw  Transportation   
2          150.0        150.0       Transportation  Transportation   
3          200.0        200.0           Embroidery            Arts   
4          400.0        400.0           Milk Sales            Food   

  country_code   country        region currency  term_in_months  lender_count  \
0           PK  Pakistan        Lahore      PKR            12.0            12   
1           PK  Pakistan        Lahore      PKR            11.0            14   
2           IN     India     Maynaguri      INR            43.0             6   
3           PK  Pakistan        Lahore      PKR            11.0             8   
4           PK  Pakistan  Abdul Hakeem      PKR            14.0            16   

  borrower_genders repayment_interval  
0           female          irregular  
1   female, female          irregular  
2           female             bullet  
3           female          irregular  
4           female            monthly

In [24]:
df.sector.unique()

array(['Food', 'Transportation', 'Arts', 'Services', 'Agriculture',
       'Manufacturing', 'Wholesale', 'Retail', 'Clothing', 'Construction',
       'Health', 'Education', 'Personal Use', 'Housing', 'Entertainment'],
      dtype=object)

In [25]:
df

funded_amount  loan_amount             activity          sector  \
0               300.0        300.0  Fruits & Vegetables            Food   
1               575.0        575.0             Rickshaw  Transportation   
2               150.0        150.0       Transportation  Transportation   
3               200.0        200.0           Embroidery            Arts   
4               400.0        400.0           Milk Sales            Food   
...               ...          ...                  ...             ...   
671199            0.0         25.0            Livestock     Agriculture   
671200            0.0         25.0            Livestock     Agriculture   
671201           25.0         25.0            Livestock     Agriculture   
671203            0.0         25.0            Livestock     Agriculture   
671204            0.0         25.0            Livestock     Agriculture   

       country_code   country        region currency  term_in_months  \
0                PK  Pakistan        Lahore      PKR            12.0   
1                PK  Pakistan        Lahore      PKR            11.0   
2                IN     India     Maynaguri      INR            43.0   
3                PK  Pakistan        Lahore      PKR            11.0   
4                PK  Pakistan  Abdul Hakeem      PKR            14.0   
...             ...       ...           ...      ...             ...   
671199           PY  Paraguay    Concepción      USD            13.0   
671200           PY  Paraguay    Concepción      USD            13.0   
671201           KE     Kenya           NaN      KES            13.0   
671203           KE     Kenya           NaN      KES            13.0   
671204           KE     Kenya           NaN      KES            13.0   

        lender_count borrower_genders repayment_interval  
0                 12           female          irregular  
1                 14   female, female          irregular  
2                  6           female             bullet  
3                  8           female          irregular  
4                 16           female            monthly  
...              ...              ...                ...  
671199             0           female            monthly  
671200             0           female            monthly  
671201             1           female            monthly  
671203             0           female            monthly  
671204             0           female            monthly  

[666984 rows x 12 columns]

### Synonyme suchen

In [26]:
df.dtypes

funded_amount         float64
loan_amount           float64
activity               object
sector                 object
country_code           object
country                object
region                 object
currency               object
term_in_months        float64
lender_count            int64
borrower_genders       object
repayment_interval     object
dtype: object

Anhand der for-Schleife werden in den object Spalten nach Synonymen geschaut 

In [27]:
lis=['activity', 'sector', 'country_code', 'country', 'currency', 'borrower_genders', 'repayment_interval']
s=[]
for a in lis:
    for b in df[a]:
        if b=='?':
            s.append(b.index)
        elif b==' ':
            s.append(b.index)
        elif b=='...':
            s.append(b.index)
        elif b=='…':
            s.append(b.index)
        elif b=='missing':
            s.append(b.index)
        elif b=='dummy':
            s.append(b.index)
        elif b=='null':
            s.append(b.index)
s          

[]

Keine weiteren Missings gefunden!

## Ausreißer 
Auf eine Betrachtung von Ausreißern wird hier verzichtet, da es hier um individuelle geschäftliche und Finanzierungstechnische Modalitäten des Individuums geht.

## Feature

Anzahl und Geschlechtertrennung 

In [28]:
temp=[]
amount=[]
male_count=[]
female_count=[]
divers_count=[]
count=0
m=0
f=0
d=0

for a in df.borrower_genders:
     temp=a.split(', ')
     for b in temp:
          count+=1
          if b=='male':
              m+=1
          elif b=='female':
              f+=1
          else:
              d+=1
     amount.append(count)
     male_count.append(m)
     female_count.append(f)
     divers_count.append(d)
     count=0
     m=0
     f=0
     d=0

amount

[1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 3,
 1,
 1,
 1,
 1,
 2,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 8,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 5,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 4,
 1,
 1,
 9,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 5,
 1,
 1,
 2,
 1,
 4,
 1,
 1,
 1,
 1,
 1,
 1,
 13,
 1,
 1,
 13,
 1,
 1,
 1,
 1,
 12,
 1,
 1,
 1,
 4,
 1,
 1,
 1,
 4,
 12,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 5,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 10,
 1,
 1,
 5,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 5,
 1,
 1,
 6,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 10,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 7,
 1,
 1,
 1,
 1,
 6,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 3,
 10,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 4,
 1,
 1,
 1,
 1,
 1,
 1,
 10,
 1,
 1,
 1,
 1,
 4,
 1,
 9,
 1,
 1,
 1,
 2,
 1,
 1,
 9,
 1,
 1,
 6,
 1,
 1,
 1,
 1,
 1,
 11,
 1,
 1,
 1,
 2,

In [29]:
df.insert(len(df.columns),'pers_count', (amount))
df.insert(len(df.columns),'male_count', (male_count))
df.insert(len(df.columns),'female_count', (female_count))

In [30]:
df

funded_amount  loan_amount             activity          sector  \
0               300.0        300.0  Fruits & Vegetables            Food   
1               575.0        575.0             Rickshaw  Transportation   
2               150.0        150.0       Transportation  Transportation   
3               200.0        200.0           Embroidery            Arts   
4               400.0        400.0           Milk Sales            Food   
...               ...          ...                  ...             ...   
671199            0.0         25.0            Livestock     Agriculture   
671200            0.0         25.0            Livestock     Agriculture   
671201           25.0         25.0            Livestock     Agriculture   
671203            0.0         25.0            Livestock     Agriculture   
671204            0.0         25.0            Livestock     Agriculture   

       country_code   country        region currency  term_in_months  \
0                PK  Pakistan        Lahore      PKR            12.0   
1                PK  Pakistan        Lahore      PKR            11.0   
2                IN     India     Maynaguri      INR            43.0   
3                PK  Pakistan        Lahore      PKR            11.0   
4                PK  Pakistan  Abdul Hakeem      PKR            14.0   
...             ...       ...           ...      ...             ...   
671199           PY  Paraguay    Concepción      USD            13.0   
671200           PY  Paraguay    Concepción      USD            13.0   
671201           KE     Kenya           NaN      KES            13.0   
671203           KE     Kenya           NaN      KES            13.0   
671204           KE     Kenya           NaN      KES            13.0   

        lender_count borrower_genders repayment_interval  pers_count  \
0                 12           female          irregular           1   
1                 14   female, female          irregular           2   
2                  6           female             bullet           1   
3                  8           female          irregular           1   
4                 16           female            monthly           1   
...              ...              ...                ...         ...   
671199             0           female            monthly           1   
671200             0           female            monthly           1   
671201             1           female            monthly           1   
671203             0           female            monthly           1   
671204             0           female            monthly           1   

        male_count  female_count  
0                0             1  
1                0             2  
2                0             1  
3                0             1  
4                0             1  
...            ...           ...  
671199           0             1  
671200           0             1  
671201           0             1  
671203           0             1  
671204           0             1  

[666984 rows x 15 columns]

In [31]:
#Kontrolle
for i in df.columns:
    print(i, df[i].nunique())

funded_amount 599
loan_amount 476
activity 163
sector 15
country_code 87
country 87
region 12695
currency 67
term_in_months 148
lender_count 499
borrower_genders 11298
repayment_interval 4
pers_count 50
male_count 40
female_count 49


Anhand den Ländern das Feature Kontinente erstellen

In [32]:
Africa = ['Africa','Kenya', 'Cote D\'Ivoire', 'Zambia','Zimbabwe','Namibia','Senegal','Tanzania','Burundi','Rwanda','Sierra Leone','Uganda','Benin','Ghana','Mozambique','Burkina Faso','Togo', 'South Africa','Malawi','Nigeria','Liberia','Mali','Madagascar','The Democratic Republic of the Congo','Congo','Cameroon','South Sudan','Somalia','Mauritania','Egypt','Lesotho']
SouthAmerica = ['South America','Peru', 'Colombia', 'Bolivia','Ecuador','Paraguay','Brazil','Suriname','Chile']
Asia=['Asia','Iraq','Indonesia','Timor-Leste','Turkey','Philippines','Palestine','Tajikistan','Jordan','Yemen','Lebanon','Israel','Azerbaijan','Mongolia','Cambodia','Armenia','Pakistan','Vietnam','Myanmar (Burma)','Kyrgyzstan','India','China','Bhutan','Nepal','Thailand','Lao People\'s Democratic Republic','Afghanistan']
NorthAmerica = ['North America','United States','Mexico','Haiti','Dominican Republic','Panama','Honduras','Nicaragua','Guatemala','El Salvador','Virgin Islands','Costa Rica','Puerto Rico','Belize','Saint Vincent and the Grenadines']
Europe=['Europe','Kosovo','Georgia','Albania','Moldova','Ukraine']
Ozeanien=['Oceania','Samoa','Guam','Vanuatu','Solomon Islands']
continents= [Africa, SouthAmerica, Asia, NorthAmerica, Europe, Ozeanien]

In [33]:
check=[]
i=0

for data in df.country:
    for kontinente in continents:
        for land in kontinente:
            if data==land:
                i=kontinente[0]
                check.append(i)
check


['Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Africa',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Africa',
 'Asia',
 'North America',
 'North America',
 'Africa',
 'Asia',
 'Asia',
 'South America',
 'Africa',
 'Asia',
 'Asia',
 'Africa',
 'Africa',
 'Africa',
 'Asia',
 'Africa',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Africa',
 'Africa',
 'Africa',
 'South America',
 'Asia',
 'Asia',
 'Asia',
 'Africa',
 'Africa',
 'Asia',
 'Africa',
 'Asia',
 'Africa',
 'Africa',
 'North America',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Asia',
 'Africa',
 'Africa',
 'North America',
 'Asia',
 'Asia',
 'Africa',
 'Africa',
 'Africa',
 'Asia',
 'Africa',
 'Africa',
 'North A

In [34]:
df['continent']=check
df

funded_amount  loan_amount             activity          sector  \
0               300.0        300.0  Fruits & Vegetables            Food   
1               575.0        575.0             Rickshaw  Transportation   
2               150.0        150.0       Transportation  Transportation   
3               200.0        200.0           Embroidery            Arts   
4               400.0        400.0           Milk Sales            Food   
...               ...          ...                  ...             ...   
671199            0.0         25.0            Livestock     Agriculture   
671200            0.0         25.0            Livestock     Agriculture   
671201           25.0         25.0            Livestock     Agriculture   
671203            0.0         25.0            Livestock     Agriculture   
671204            0.0         25.0            Livestock     Agriculture   

       country_code   country        region currency  term_in_months  \
0                PK  Pakistan        Lahore      PKR            12.0   
1                PK  Pakistan        Lahore      PKR            11.0   
2                IN     India     Maynaguri      INR            43.0   
3                PK  Pakistan        Lahore      PKR            11.0   
4                PK  Pakistan  Abdul Hakeem      PKR            14.0   
...             ...       ...           ...      ...             ...   
671199           PY  Paraguay    Concepción      USD            13.0   
671200           PY  Paraguay    Concepción      USD            13.0   
671201           KE     Kenya           NaN      KES            13.0   
671203           KE     Kenya           NaN      KES            13.0   
671204           KE     Kenya           NaN      KES            13.0   

        lender_count borrower_genders repayment_interval  pers_count  \
0                 12           female          irregular           1   
1                 14   female, female          irregular           2   
2                  6           female             bullet           1   
3                  8           female          irregular           1   
4                 16           female            monthly           1   
...              ...              ...                ...         ...   
671199             0           female            monthly           1   
671200             0           female            monthly           1   
671201             1           female            monthly           1   
671203             0           female            monthly           1   
671204             0           female            monthly           1   

        male_count  female_count      continent  
0                0             1           Asia  
1                0             2           Asia  
2                0             1           Asia  
3                0             1           Asia  
4                0             1           Asia  
...            ...           ...            ...  
671199           0             1  South America  
671200           0             1  South America  
671201           0             1         Africa  
671203           0             1         Africa  
671204           0             1         Africa  

[666984 rows x 16 columns]

In [35]:
#Feature Auszahlung pro Monat
df.insert(len(df.columns),'pay_monat', (df.funded_amount/df.term_in_months))

In [36]:
#Feature Auszahlung pro Monat und Person
df.insert(len(df.columns),'pay_mp', (df.funded_amount/df.term_in_months/df.pers_count))

In [37]:
#Feature Zielbetrag erreicht
df.insert(len(df.columns),'goal_pctg', (100/df.loan_amount*df.funded_amount))

In [38]:
#Feature Funding amount pro Person
df.insert(len(df.columns),'funded_pp', (df.funded_amount/df.pers_count))

In [39]:
#Runden 
df.loc[:,'pay_mp']=round(df.pay_mp,2)

In [40]:
df.loc[:,'pay_monat']=round(df.pay_monat,2)

## Speicherplatzreduzierung

In [41]:
# Betrachtung des Speichers
print((df.memory_usage(deep=True).sum()/1000000), "MB")
df.memory_usage(deep=True)

455.095986 MB


Index                  5335872
funded_amount          5335872
loan_amount            5335872
activity              46228811
sector                43330076
country_code          39352056
country               43649696
region                45943060
currency              40019040
term_in_months         5335872
lender_count           5335872
borrower_genders      46904008
repayment_interval    43129882
pers_count             5335872
male_count             5335872
female_count           5335872
continent             42508893
pay_monat              5335872
pay_mp                 5335872
goal_pctg              5335872
funded_pp              5335872
dtype: int64

In [42]:
# Betrachtung von Datentypen unnd Anzahl
df.dtypes 

funded_amount         float64
loan_amount           float64
activity               object
sector                 object
country_code           object
country                object
region                 object
currency               object
term_in_months        float64
lender_count            int64
borrower_genders       object
repayment_interval     object
pers_count              int64
male_count              int64
female_count            int64
continent              object
pay_monat             float64
pay_mp                float64
goal_pctg             float64
funded_pp             float64
dtype: object

In [43]:
df_opti=df.copy()
df_opti.memory_usage(deep=True) 

Index                  5335872
funded_amount          5335872
loan_amount            5335872
activity              46228811
sector                43330076
country_code          39352056
country               43649696
region                45943060
currency              40019040
term_in_months         5335872
lender_count           5335872
borrower_genders      46904008
repayment_interval    43129882
pers_count             5335872
male_count             5335872
female_count           5335872
continent             42508893
pay_monat              5335872
pay_mp                 5335872
goal_pctg              5335872
funded_pp              5335872
dtype: int64

In [44]:
# Wo können floats in ints umgewandelt werden?

#Schleife kontrolliert und speichert die columns von float dtypes
lis=[]
for i in df_opti.columns:
    if df_opti[i].dtype=='float64':
        lis.append(i)
        
#Schleife zählt anzahl der einträge bei denen keine nachkommastelle auftritt        
#die ersten drei 
a=0
for i in lis:
    for j in df_opti[i]:
        if j%1==0:
            a+=1
            
    print(j, a)
    a=0

0.0 666984
25.0 666984
13.0 666984
0.0 98027
0.0 86648
0.0 623825
0.0 602332


In [45]:
# in den ersten drei einträgen von lis tritt keine kommazahl auf.
# Wir können diese problemlos umformen.
df_opti[lis[0]]=df_opti[lis[0]].astype(int)
df_opti[lis[1]]=df_opti[lis[1]].astype(int)
df_opti[lis[2]]=df_opti[lis[2]].astype(int)

df_opti.dtypes

funded_amount           int32
loan_amount             int32
activity               object
sector                 object
country_code           object
country                object
region                 object
currency               object
term_in_months          int32
lender_count            int64
borrower_genders       object
repayment_interval     object
pers_count              int64
male_count              int64
female_count            int64
continent              object
pay_monat             float64
pay_mp                float64
goal_pctg             float64
funded_pp             float64
dtype: object

In [46]:
#Umwandlung von Strings in Categories

for i in df_opti.columns:
    print(i, df_opti[i].nunique())

funded_amount 599
loan_amount 476
activity 163
sector 15
country_code 87
country 87
region 12695
currency 67
term_in_months 148
lender_count 499
borrower_genders 11298
repayment_interval 4
pers_count 50
male_count 40
female_count 49
continent 6
pay_monat 5750
pay_mp 8367
goal_pctg 4965
funded_pp 5188


In [47]:

lis=['activity','sector','country_code','country',"region","currency","borrower_genders","continent",'repayment_interval']

for i in lis:
    df_opti.loc[:,i] = df_opti.loc[:,i].astype("category")

df_opti.dtypes

funded_amount            int32
loan_amount              int32
activity              category
sector                category
country_code          category
country               category
region                category
currency              category
term_in_months           int32
lender_count             int64
borrower_genders      category
repayment_interval    category
pers_count               int64
male_count               int64
female_count             int64
continent             category
pay_monat              float64
pay_mp                 float64
goal_pctg              float64
funded_pp              float64
dtype: object

In [48]:
reduction = (df.memory_usage(deep=True).sum()-df_opti.memory_usage(deep=True).sum())/df.memory_usage(deep=True).sum()

print(f"{reduction:0f}")

0.850994


In [49]:
print((df_opti.memory_usage(deep=True).sum()/1000000), "MB")
df_opti.memory_usage(deep=True)
data0=df_opti.memory_usage(deep=True).sum()/1000000
data2=df.memory_usage(deep=True).sum()/1000000
print('Die Daten wurden um ', data2-data0, 'MB eingekürzt')
print('Allein im zweiten Speicheroptimierungs-Schritt wurden ', data2-data0, 'MB eingespart')
print('Ich habe den Speicher seit ln11 um insgesamt ', round(100-(100/data2*data0),2), '% eingespart')

67.812195 MB
Die Daten wurden um  387.28379099999995 MB eingekürzt
Allein im zweiten Speicheroptimierungs-Schritt wurden  387.28379099999995 MB eingespart
Ich habe den Speicher seit ln11 um insgesamt  85.1 % eingespart


In [50]:
df_opti

funded_amount  loan_amount             activity          sector  \
0                 300          300  Fruits & Vegetables            Food   
1                 575          575             Rickshaw  Transportation   
2                 150          150       Transportation  Transportation   
3                 200          200           Embroidery            Arts   
4                 400          400           Milk Sales            Food   
...               ...          ...                  ...             ...   
671199              0           25            Livestock     Agriculture   
671200              0           25            Livestock     Agriculture   
671201             25           25            Livestock     Agriculture   
671203              0           25            Livestock     Agriculture   
671204              0           25            Livestock     Agriculture   

       country_code   country        region currency  term_in_months  \
0                PK  Pakistan        Lahore      PKR              12   
1                PK  Pakistan        Lahore      PKR              11   
2                IN     India     Maynaguri      INR              43   
3                PK  Pakistan        Lahore      PKR              11   
4                PK  Pakistan  Abdul Hakeem      PKR              14   
...             ...       ...           ...      ...             ...   
671199           PY  Paraguay    Concepción      USD              13   
671200           PY  Paraguay    Concepción      USD              13   
671201           KE     Kenya           NaN      KES              13   
671203           KE     Kenya           NaN      KES              13   
671204           KE     Kenya           NaN      KES              13   

        lender_count borrower_genders repayment_interval  pers_count  \
0                 12           female          irregular           1   
1                 14   female, female          irregular           2   
2                  6           female             bullet           1   
3                  8           female          irregular           1   
4                 16           female            monthly           1   
...              ...              ...                ...         ...   
671199             0           female            monthly           1   
671200             0           female            monthly           1   
671201             1           female            monthly           1   
671203             0           female            monthly           1   
671204             0           female            monthly           1   

        male_count  female_count      continent  pay_monat  pay_mp  goal_pctg  \
0                0             1           Asia      25.00   25.00      100.0   
1                0             2           Asia      52.27   26.14      100.0   
2                0             1           Asia       3.49    3.49      100.0   
3                0             1           Asia      18.18   18.18      100.0   
4                0             1           Asia      28.57   28.57      100.0   
...            ...           ...            ...        ...     ...        ...   
671199           0             1  South America       0.00    0.00        0.0   
671200           0             1  South America       0.00    0.00        0.0   
671201           0             1         Africa       1.92    1.92      100.0   
671203           0             1         Africa       0.00    0.00        0.0   
671204           0             1         Africa       0.00    0.00        0.0   

        funded_pp  
0           300.0  
1           287.5  
2           150.0  
3           200.0  
4           400.0  
...           ...  
671199        0.0  
671200        0.0  
671201       25.0  
671203        0.0  
671204        0.0  

[666984 rows x 20 columns]

## Speichern

In [51]:
df=df_opti
df.to_pickle('df_after_processing.pkl')